tensorboard/data/server/gcs/auth.rs

```diff
@@ -0,0 +1,298 @@
+    ///
+    /// This operation is cheap and does not actually fetch any access tokens.
+    pub fn new(creds: Credentials) -> Self {
+        Self {
+            creds,
+            token: RwLock::new(None),
+        }
+    }
+}
+
+/// An access token that's valid until a particular point in time.
+#[derive(Debug)]
+struct BoundedToken {
+    access_token: AccessToken,
+    expires: Instant,
+}
+
+impl BoundedToken {
+    /// Tests whether this token will still be valid at the given instant.
+    fn valid_at(&self, when: Instant) -> bool {
+        when < self.expires
+    }
+}
+
+/// Private access token module to prevent accidental leaking of tokens into logs, etc. An access
+/// token can be attached to a request, but cannot be directly extracted.
+mod access_token {
+    use super::*;
+    #[derive(Deserialize)]
+    pub struct AccessToken(String);
+    impl AccessToken {
+        /// Attaches this token to an outgoing request.
+        pub fn authenticate(&self, rb: RequestBuilder) -> RequestBuilder {
+            rb.bearer_auth(&self.0)
+        }
+    }
+    impl Debug for AccessToken {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            f.debug_tuple("AccessToken")
+                .field(&format_args!("_"))
+                .finish()
+        }
+    }
+}
+use access_token::AccessToken;
+
+impl TokenStore {
+    /// Attempts to attach an access token to the given outgoing request.
+    ///
+    /// A cached token will be reused if it is expected to be valid for at least `lifetime`.
+    /// Otherwise, a new token will be fetched and stored.
+    pub fn authenticate(
+        &self,
+        rb: RequestBuilder,
+        http: &HttpClient,
+        lifetime: Duration,
+    ) -> RequestBuilder {
+        if self.creds.anonymous() {
+            return rb;
+        }
+        let token = self.token.read().expect("failed to read auth token");
+        if let Some(t) = Self::check_expiry(&*token, lifetime) {
+            return t.authenticate(rb);
+        }
+        drop(token);
+        let mut token = self.token.write().expect("failed to write auth token");
+        // Check again: may have just been written by a different client, in which case no need to
+        // re-fetch.
+        if let Some(t) = Self::check_expiry(&*token, lifetime) {
+            return t.authenticate(rb);
+        }
+        // If we get here, we need a fresh token.
+        *token = self.creds.fetch(http);
+        if let Some(ref t) = *token {
+            debug!(
+                "Obtained new access token, live for the next {:?}",
+                t.expires.saturating_duration_since(Instant::now())
+            );
+            t.access_token.authenticate(rb)
+        } else {
+            rb
+        }
+    }
+
+    /// Checks whether `token` represents a token that will still be valid for at least the given
+    /// `lifetime`, and if so returns a reference to the inner access token.
```

Comment:
Slightly non-obvious IMO from the name that `check_expiry` also unwraps the token.  Optional but maybe consider making it a method on BoundedToken (could replace `valid_at`) like `unwrap_if_valid_for()`?

Reply:
Okay, sure: changed as suggested.